source.go

```diff
@@ -408,3 +411,64 @@ func PatchSources(worker llb.State, spec *Spec, sourceToState map[string]llb.Sta
 
 	return states
 }
+
+// Tar creates a tar+gz from the provided state and puts it in the provided dest.
+// The provided work state is used to perform the neccessary operations to produce the tarball and requires the tar and gzip binaries.
+func Tar(work, src llb.State, dest string, opts ...llb.ConstraintsOpt) llb.State {
+	// Put the output tar in a consistent location regardless of `dest`
+	// This way if `dest` changes we don't have to rebuild the tarball, which can be expensive.
+	outBase := "/tmp/out"
+	out := filepath.Join(outBase, filepath.Dir(dest))
+	worker := work.Run(
+		llb.AddMount("/src", src, llb.Readonly),
+		llb.Args([]string{"/bin/sh", "-c", "tar -C /src -cvzf /tmp/st ."}),
+		WithConstraints(opts...),
+	).
+		Run(
+			llb.Args([]string{"/bin/sh", "-c", "mkdir -p " + out + " && mv /tmp/st " + filepath.Join(out, filepath.Base(dest))}),
+			WithConstraints(opts...),
+		)
+
+	return worker.AddMount(outBase, llb.Scratch())
+}
+
+// SourceToTar provides a common SourceModifier that will produce a tarball from the provided source.
+// The tarball will be called <key>.tar.gz
+// The input source state is discarded and the tarball is returned.
+func SourceToTar(work llb.State) SourceModifier {
+	return func(key string, src *Source, st llb.State, opts ...llb.ConstraintsOpt) (llb.State, error) {
+		if isDir, err := SourceIsDir(*src); !isDir || err != nil {
+			return st, err
+		}
+		return Tar(work, st, key+".tar.gz", opts...), nil
+	}
+}
+
+// SourceModifier is a function that modifies the provided source LLB state.
+type SourceModifier func(key string, src *Source, st llb.State, opts ...llb.ConstraintsOpt) (llb.State, error)
+
+// Sources gets all the source LLB states from the spec.
+//
+// SourceModifiers can be provided to modify the source LLB state before it is returned.
+// The `key` argument to the SourceModifier is the name of the source.
+func Sources(spec *Spec, sOpt SourceOpts, mods ...SourceModifier) (map[string]llb.State, error) {
```

Comment:
I like this a lot. It is very clean with the passed in modifiers, and the returned map prevents us from having to construct the name -> source state mapping later on